jsonschema/src/keywords/unevaluated_properties.rs

```diff
@@ -0,0 +1,1090 @@
+                had_match = true;
+
+                let pattern_output = node.apply_rooted(property_instance, property_path);
+                output += pattern_output;
+            }
+        }
+
+        had_match.then(|| output)
+    }
+}
+
+/// A subvalidator for subschema validation such as `allOf`, `oneOf`, and `anyOf`.
+///
+/// Unlike the validation logic for `allOf`/`oneOf`/`anyOf` themselves, this subvalidator searches
+/// configured subvalidators in a first-match-wins process. For example, a property will be
+/// considered evaluated against subschemas defined via `oneOf` so long as one subschema would evaluate
+/// the property, even if, say, more than one subschema in `oneOf` is technically valid, which would
+/// otherwise be a failure for validation of `oneOf` in and of itself.
+#[derive(Debug)]
+struct SubschemaSubvalidator {
+    subvalidators: Vec<UnevaluatedPropertiesValidator>,
+}
+
+impl SubschemaSubvalidator {
+    fn is_valid_property(
+        &self,
+        instance: &Value,
+        property_instance: &Value,
+        property_name: &str,
+    ) -> Option<bool> {
+        self.subvalidators.iter().find_map(|subvalidator| {
+            subvalidator.is_valid_property(instance, property_instance, property_name)
+        })
+    }
+
+    fn validate_property<'instance>(
+        &self,
+        instance: &'instance Value,
+        instance_path: &InstancePath,
+        property_path: &InstancePath,
+        property_instance: &'instance Value,
+        property_name: &str,
+    ) -> Option<ErrorIterator<'instance>> {
+        self.subvalidators.iter().find_map(|subvalidator| {
+            subvalidator.validate_property(
+                instance,
+                instance_path,
+                property_path,
+                property_instance,
+                property_name,
+            )
+        })
+    }
+
+    fn apply_property<'a>(
+        &'a self,
+        instance: &Value,
+        instance_path: &InstancePath,
+        property_path: &InstancePath,
+        property_instance: &Value,
+        property_name: &str,
+    ) -> Option<BasicOutput<'a>> {
+        self.subvalidators.iter().find_map(|subvalidator| {
+            subvalidator.apply_property(
+                instance,
+                instance_path,
+                property_path,
+                property_instance,
+                property_name,
+            )
+        })
+    }
+}
+
+/// A subvalidator for unevaluated properties.
+#[derive(Debug)]
+struct UnevaluatedSubvalidator {
+    // The validator associated with `unevaluatedProperties`.
+    //
+    // When `None`, this indicates that `unevaluatedProperties` was `false`, and that no unevaluated
+    // properties are allowed at all. If `Some(None)`, then unevaluated properties _are_ allowed,
+    // and if `Some(Some(...))`, then the contained validator is used to validate any unevaluated
+    // properties.
+    node: Option<Option<SchemaNode>>,
```

Comment:
I am thinking if it would make sense to have a separate enum where each variant will denote each case, instead of using double `Option`. What do you think?

Reply:
That's fair.

I had sort of the same thought, especially when I was writing the doc comment to explain the behavior of the double `Option<T>`... so I probably should have taken that as a hint in the moment that it should be an enum. :)

Reply:
I addressed this in a37e8ca.